### ISSUE TITLE
CSVFeedSpider has a huge memory usage when dealing with a csv file contains huge amount of urls

### ISSUE THREAD

**Issue by HaujetZhao** [scrapy/scrapy#5330]
### Description

CSVFeedSpider has a huge memory usage when dealing with a csv file contains huge amount of urls

### Steps to Reproduce

1. Prepare a csv file contains 1000 urls, each row contains one url
2. Use CSVFeedSpider to follow each url in `parse_row`
3. Watch the memory usage, it's normal. Nothing too strange
4. Prepare a csv file contains 200000 urls, each row contains one url
5. Use CSVFeedSpider to follow (`response.follow`) each url in `parse_row`
6. Watch the memory usage, you will notice: 
    - The memory usage increased to about 1GB
    - Then the Requests begin to download pages

**Expected behavior:** [What you expect to happen]

The Requests number should be equal to CONCURRENT_REQUESTS, then iterate the remained rows as the downloading goes, not create Requests for all rows at first. 

**Actual behavior:** [What actually happens]

So, it seems the CSVFeedSpider first walk through all of the csv rows and created Requests for every url, then it begin to download. 

**Reproduces how often:** [What percentage of the time does it reproduce?]

100%

### Versions

Scrapy       : 2.5.1
lxml         : 4.6.3.0
libxml2      : 2.9.5
cssselect    : 1.1.0
parsel       : 1.6.0
w3lib        : 1.22.0
Twisted      : 21.7.0
Python       : 3.8.6 (tags/v3.8.6:db45529, Sep 23 2020, 15:52:53) [MSC v.1927 64 bit (AMD64)]
pyOpenSSL    : 21.0.0 (OpenSSL 1.1.1i  8 Dec 2020)
cryptography : 3.3.1
Platform     : Windows-10-10.0.22000-SP0

**Comment by wRAR** [comment]
`CSVFeedSpider.parse_rows` is just a callback and there is currently no way to pause request creation from a callback, even more so pause until the queue is empty/not full. When you emit 200000 requests from a callback they are always created immediately, this is not specific to `CSVFeedSpider`. 

OTOH requests emitted from `start_requests` (note that you don't need `CSVFeedSpider` when the file is local, you can open it directly in `start_requests`) actually wait until the queue is empty, though this may not be what you want if you also create requests from other callbacks.

**Comment by HaujetZhao** [comment]
I rewrote it with `Threading.Thread` which can observe the queue and enqueue in a controlled pace. 

In the documentation, there is few words about accessing the queue from a Spider or Middleware. I think this can be improved.

**Comment by wRAR** [comment]
Indeed, I don't think it was ever recommended to access the queues directly.

**Comment by HaujetZhao** [comment]
Then I think Scrapy should natively provide a way to controll the speed of generating Requests.So this should be a Feature Request. 

In this case, one csv Response returned 200000 urls, when the Requests in the queue are used, it's been a long time, the cookie or headers or proxy may be ununsable. 

My suggesting is: when the Spider follow a link: 

```python
def parse_row(self, response, row):
        return response.follow(row['url'], callback=self.parse)
```
do not generate Requests and inqueue at once. Instead, put the links in a SQLite database stored on disk, use another Thread to watch the queue in a for loop, when the queue number is less than some number (say `2 * concurrent_requests`), fetch a url from database, generate a new Request, inqueue.

**Comment by wRAR** [comment]
> do not generate Requests and inqueue at once. Instead, put the links in a SQLite database stored on disk, use another Thread to watch the queue in a for loop, when the queue number is less than some number (say 2 * concurrent_requests), fetch a url from database, generate a new Request, inqueue.

This is definitely not something we want to implement, especially by default. It may be possible to make something fitting the Scrapy architecture that addresses your requirements, though I don't see how without spending some time thinking about it. 

We had an idea about providing some object that you can await on in your callback or `start_requests` method to know when the queue becomes empty or not full, maybe it will be implemented some day.

**Comment by Gallaecio** [comment]
If the issue is high memory usage when the scheduler has a lot of requests, [`JOBDIR`](https://docs.scrapy.org/en/latest/topics/jobs.html) may be the way to go with the current implementation.

**Comment by Gallaecio** [comment]
Should we close this?

**Comment by AdrianAtZyte** [comment]
https://github.com/scrapy/scrapy/pull/7883 clarifies docs on how to pace request creation, I think we can consider this closed with those documentation improvements.
